extensions/default/src/MergeDataSource/index.js

```diff
@@ -0,0 +1,195 @@
+ */
+const getDedupedData = ({ globalData, incomingData, mergeKey, level }) => {
+  const keys = globalData.map(r => getKeyByLevel(r, mergeKey, level));
+  return incomingData.filter(r => !keys.includes(getKeyByLevel(r, mergeKey, level)));
+};
+
+/**
+ * Processes deduplicated data and pushes it to globalData array.
+ * @param {Object} options - The options object.
+ * @param {string} options.sourceName - The name of the data source.
+ * @param {Array} options.globalData - The global data array.
+ * @param {Array} options.incomingData - The incoming data array.
+ * @param {string} options.mergeKey - The key to merge the data on.
+ * @param {number} options.level - The level to merge the data on.
+ * @param {Function} options.tagFunc - The function to tag the data with.
+ */
+const processDedupedData = ({ sourceName, globalData, incomingData, mergeKey, level, tagFunc }) => {
+  let data = [];
+  if (mergeKey && level !== -1) {
+    const dedupedData = getDedupedData({
+      globalData: globalData.flat(),
+      incomingData,
+      mergeKey,
+      level,
+    });
+    data = data.concat(dedupedData);
+  } else {
+    data = data.concat(incomingData);
+  }
+  globalData.push(tagFunc(data, sourceName));
+};
+
+function createMergeDataSourceApi(mergeConfig, UserAuthenticationService, extensionManager) {
+  const { seriesMerge } = mergeConfig;
+  const { dataSourceNames } = seriesMerge;
+
+  /**
+   * Calls a specified function on all data sources and merges the results.
+   *
+   * @async
+   * @function
+   * @param {string} path - The path of the function to call.
+   * @param {Object} args - The arguments to pass to the function.
+   * @returns {Promise<Array>} - A promise that resolves to an array of merged data.
+   */
+  const callForAllDataSourcesAsync = async (path, args) => {
+    const mergeMap = {
+      'query.studies.search': {
+        mergeKey: 'studyInstanceUid',
+        level: 0,
+        /** TODO */
+        tagFunc: x => x,
+      },
+      'retrieve.series.metadata': {
+        mergeKey: 'seriesInstanceUid',
+        level: 0,
+        tagFunc: (series, sourceName) => {
+          series.forEach(series => {
+            series.RetrieveAETitle = sourceName;
+            DicomMetadataStore.updateSeriesMetadata(series);
+          });
+          return series;
+        },
+      },
+    };
+
+    const { mergeKey, level, tagFunc } = mergeMap[path] || { tagFunc: x => x, level: -1 };
+
+    const dataSourceDefs = Object.values(extensionManager.dataSourceDefs);
+    const promises = [];
+    let globalData = [];
+
+    for (const dataSourceDef of dataSourceDefs) {
+      const { configuration, sourceName } = dataSourceDef;
+      if (configuration && dataSourceNames.includes(sourceName)) {
+        const [dataSource] = extensionManager.getDataSources(sourceName);
+        const promise = executeFunction(dataSource, path, args);
+        promises.push(promise);
+
+        promise.then(data =>
```

Comment:
You need to promises.push(promise.then(data => ....
Otherwise you have introduced a very subtle bug when the allSettled can be complete, but the merge is not, and the length ends up changing after the response is returned.

Reply:
Updated, thanks!